ultralytics/engine/results.py

```diff
@@ -327,6 +334,102 @@ def verbose(self):
                 log_string += f"{n} {self.names[int(c)]}{'s' * (n > 1)}, "
         return log_string
 
+    def asdict(self, py_typed: bool = True) -> dict:
+        """
+        Convert the detection results for any task into a dictionary format.
+
+        Args:
+            py_typed (bool): Whether to convert values in the dictionary to Python types; used to save as YAML.
+
+        Returns:
+            dict: A dictionary containing results data.
+        """
+        # Check task data
+        box, mask, probs, kp, obb = [getattr(self, t) is not None and any([getattr(self, t)]) for t in self._keys]
+        r = self.cpu().numpy()
+        lbl_idx = [r.probs.top1] if probs else (getattr(r.obb, "cls", None) if obb else getattr(r.boxes, "cls", None))
+        all_conf = (
+            [r.probs.top1conf] if probs else (getattr(r.obb, "conf", None) if obb else getattr(r.boxes, "conf", None))
+        )
+        all_ids = r.boxes.id if box else getattr(r.obb, "id", None)
+
+        # Generate results dictionary if not populated
+        if not self.result_dict:
+            self.result_dict = {
+                Path(self.path).name: {
+                    "detections": {
+                        n: {
+                            "label": self.names.get(lbl_idx[n] if np.array(lbl_idx).any() else None),
+                            "conf": float(all_conf[n]) if np.array(all_conf).any() else None,
+                            "id": int(all_ids[n]) if all_ids is not None else None,
+                            "xyxy": r.boxes.xyxy[n] if box else None,
+                            "nxyxy": r.boxes.xyxyn[n] if box else None,
+                            "xywh": r.boxes.xywh[n] if box else None,
+                            "nxywh": r.boxes.xywhn[n] if box else None,
+                            "mask-xy": self.masks.xy[n] if mask and box else None,  # NOTE cpu+numpy missing attr
+                            "mask-xyn": self.masks.xyn[n] if mask and box else None,  # NOTE cpu+numpy missing attr
+                            "kp-xy": r.keypoints.xy[n] if kp and box else None,
+                            "kp-xyn": r.keypoints.xyn[n] if kp and box else None,
+                            "kp-conf": r.keypoints.conf[n] if kp and box else None,
+                            "xywhr": r.obb.xywhr[n] if obb else None,
+                            "xyxyxyxy": r.obb.xyxyxyxy[n] if obb else None,
+                            "xyxyxyxyn": r.obb.xyxyxyxyn[n] if obb else None,
+                            "top1": r.probs.top1 if probs else None,
+                            "top1conf": r.probs.top1conf if probs else None,
+                            "top5": r.probs.top5 if probs else None,
+                            "top5conf": r.probs.top5conf if probs else None,
+                        }
+                        for n in range(max(len(lbl_idx or []), 1))
```

Comment:
This portion of the code presents a potential ambiguity:
```suggestion
                        if lbl_idx is None:
                            length = 1
                        else:
                            length = max(len(lbl_idx), 1)
                        for n in range(length):

```

Reply:
That's fair. I think I will abstain from making the call as to which version to use and let @glenn-jocher make the final call for this portion of the code (once we wrap up everything else).